CMakeLists.txt

```diff
@@ -90,86 +85,53 @@ option(IREE_ENABLE_EMITC "Enables MLIR EmitC dependencies." ${IREE_BUILD_COMPILE
-
-  # Vulkan is not natively supported on Apple platforms.
-  # Metal should generally be used instead, though MoltenVK may also work.
-  if(APPLE)
-    list(REMOVE_ITEM IREE_HAL_DRIVERS_TO_BUILD Vulkan)
-  endif()
-
-  # CUDA is not natively supported on Android or Apple platforms.
-  if(ANDROID OR APPLE)
-    list(REMOVE_ITEM IREE_HAL_DRIVERS_TO_BUILD CUDA)
-  endif()
+# CUDA is not natively supported on Android or Apple platforms.
+if(ANDROID OR APPLE)
+  set(IREE_HAL_DRIVER_CUDA_DEFAULT OFF)
+else()
+  set(IREE_HAL_DRIVER_CUDA_DEFAULT ON)
 endif()
-message(STATUS "Building runtime HAL drivers: ${IREE_HAL_DRIVERS_TO_BUILD}")
-
-# Start each IREE_HAL_DRIVER_* OFF, then enable from IREE_HAL_DRIVERS_TO_BUILD.
-foreach(_backend ${IREE_ALL_HAL_DRIVERS})
-  string(TOUPPER "${_backend}" uppercase_backend)
-  set(IREE_HAL_DRIVER_${uppercase_backend} OFF CACHE BOOL "" FORCE)
-endforeach()
-foreach(_backend ${IREE_HAL_DRIVERS_TO_BUILD})
-  string(TOUPPER "${_backend}" uppercase_backend)
-  string(REPLACE "\"" "" uppercase_backend ${uppercase_backend})
-  set(IREE_HAL_DRIVER_${uppercase_backend} ON CACHE BOOL "" FORCE)
-endforeach()
-
-# List of all compiler target backends included in the project:
-set(IREE_ALL_TARGET_BACKENDS
-  CUDA
-  Dylib-LLVM-AOT
-  WASM-LLVM-AOT
-  Metal-SPIRV
-  ROCm
-  Vulkan-SPIRV
-  VMVX
-  WebGPU
-)
 
-if(${IREE_BUILD_COMPILER})
-  if(IREE_TARGET_BACKENDS_TO_BUILD STREQUAL "default")
-    set(IREE_TARGET_BACKENDS_TO_BUILD ${IREE_ALL_TARGET_BACKENDS})
-
-    # Disable the WebGPU backend by default, as it has complex dependencies and
-    # is still early in its development.
-    list(REMOVE_ITEM IREE_TARGET_BACKENDS_TO_BUILD WebGPU)
-  endif()
-  message(STATUS "Building compiler target backends: ${IREE_TARGET_BACKENDS_TO_BUILD}")
+# Vulkan is not natively supported on Apple platforms.
+# Metal should generally be used instead, though MoltenVK may also work.
+if (APPLE)
+  set(IREE_HAL_DRIVER_VULKAN_DEFAULT OFF)
 else()
-  set(IREE_TARGET_BACKENDS_TO_BUILD "" CACHE STRING "" FORCE)
-  message(STATUS "Compiler is disabled, building no target backends")
+  set(IREE_HAL_DRIVER_VULKAN_DEFAULT ON)
 endif()
 
-# Default every IREE_TARGET_BACKEND_* to OFF
-foreach(_backend ${IREE_ALL_TARGET_BACKENDS})
-  string(TOUPPER "${_backend}" uppercase_backend)
-  set(IREE_TARGET_BACKEND_${uppercase_backend} OFF CACHE BOOL "" FORCE)
-endforeach()
-
-# Set IREE_TARGET_BACKEND_* based on configuration
-foreach(_backend ${IREE_TARGET_BACKENDS_TO_BUILD})
-  string(TOUPPER "${_backend}" uppercase_backend)
-  string(REPLACE "\"" "" uppercase_backend ${uppercase_backend})
-  set(IREE_TARGET_BACKEND_${uppercase_backend} ON CACHE BOOL "" FORCE)
-endforeach()
-
-list(APPEND CMAKE_MODULE_PATH
-  ${CMAKE_CURRENT_LIST_DIR}/build_tools/cmake/
-  ${CMAKE_CURRENT_LIST_DIR}/bindings/python/build_tools/cmake/
-)
+option(IREE_HAL_DRIVER_CUDA "Enables the CUDA runtime HAL driver" ${IREE_HAL_DRIVER_CUDA_DEFAULT})
```

Comment:
Whooo! This has always sketched me out also

As noted in Ben's comment, I think that having one easy variable to set things can still be nice maybe. But I think more important: it gave the ability to list *exactly* which drivers you wanted to build. If I only want Dylib, it's pretty annoying to have to list every single thing I want to disable, especially if that set changes over time. The single string variable gave that option, though it also was not super ergonomic to use. Another way to achieve that would be to have a single variable for setting the default for all "standard" drivers, which defaults to ON, then I can set it to OFF and only enable the one driver I want.

Something like

```cmake
option(IREE_HAL_DRIVER_DEFAULT "Sets the default value for all standard drivers. Is overridden by platform-specific logic (e.g. disable cuda on android) and explicitly setting the variable for a specific driver" ON)

if (APPLE)
  set(IREE_HAL_DRIVER_VULKAN_DEFAULT OFF)
else()
  set(IREE_HAL_DRIVER_VULKAN_DEFAULT ${IREE_HAL_DRIVER_DEFAULT})
endif()

option(IREE_HAL_DRIVER_DYLIB "Enables the Dylib runtime HAL driver" ${IREE_HAL_DRIVER_DEFAULT})
```

Reply:
Nice, overriding the default is pretty clean. Thanks for the suggestion!